core/chain-configs/src/config_store.rs

```diff
@@ -0,0 +1,101 @@
+use crate::genesis_config::GenesisConfigSnapshot;
+use std::fs::File;
+use std::io::Read;
+use serde_json;
+use std::env;
+
+pub const RESOURCES_DIR: &str = "core/chain-configs/res/";
+pub const CONFIG_PATCH_LIST_FILENAME: &str = "config_patch_list.json";
+
+/// Stores epochs when new config patches should be applied.
+/// This list is used to load files with config patches.
+#[derive(Debug, serde::Deserialize)]
+pub struct ConfigPatchList {
+    epoch_patches: Vec<EpochHeight>,
+}
+
+/// Stores chain config for each epoch where it was updated.
+#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+pub struct ChainConfigStore {
+    /// Initial chain config when bootstrapping network at epoch 0.
+    pub initial_chain_config: Arc<ChainConfig>,
+    /// Maps epoch to the config.
+    store: BTreeMap<EpochHeight, Arc<ChainConfigPatch>>,
+}
+
+impl ChainConfigStore {
+    /// Constructs a new store. Initial chain config is read from provided genesis_config_snapshot.
+    /// Stored values are read from config path list file and specific patches are read from the
+    /// files that are names after the epoch when they should be applied.
+    pub fn new(genesis_config_snapshot: GenesisConfigSnapshot) -> Self {
+        let initial_chain_config = Arc::new(ChainConfig::new(genesis_config_snapshot));
+        let mut store = BTreeMap::new();
+        Self::populate_config_store(&mut store);
+        Self { initial_chain_config, store }
+    }
+
+    fn populate_config_store(store: &mut BTreeMap<EpochHeight, Arc<ChainConfigPatch>>) {
+        let config_patch_list = Self::get_config_patch_list();
+
+        for epoch_height in config_patch_list.epoch_patches {
+            let chain_config_patch = Self::load_chain_config(epoch_height);
+            store.insert(epoch_height, Arc::new(chain_config_patch));
+        }
+    }
+
+    fn load_chain_config(epoch_height: EpochHeight) -> ChainConfigPatch {
+        let current_dir = env::current_dir().expect("Failed to get the current directory");
+        let path = current_dir.join(RESOURCES_DIR).join(epoch_height.to_string() + ".json");
+        let mut file = File::open(&path).expect("Failed to open chain config patch file.");
+        let mut json_str = String::new();
+        file.read_to_string(&mut json_str)
+            .expect("Failed to read the chain config patch file to string. ");
+        let json_str_without_comments: String =
+            near_config_utils::strip_comments_from_json_str(&json_str)
+                .expect("Failed to strip comments from chain config patch file.");
+        serde_json::from_str(&json_str_without_comments)
+            .expect("Failed to deserialize the chain config patch file.")
+    }
+
+    fn get_config_patch_list() -> ConfigPatchList {
+        let current_dir = env::current_dir().expect("Failed to get the current directory");
+        let path = current_dir.join(RESOURCES_DIR).join(CONFIG_PATCH_LIST_FILENAME);
```

Comment:
Sure, I will add this to the CLI. I will update everything in the batch next week and push it to the PR with all other agreed changes.

@jakmeier regarding this is it a good approach to have a file with all the changes (based on Epoch Id or Protocol Version) and a file per version like here? Or in other words, is it ok to keep everything the same, just pass the path of the changes file via CLI?

Reply:
@jakmeier regarding this, I started implementing it with the CLI and the code looks much more dirty. Do you maybe know why the Runtime store does not have the CLI option? It also has hardcoded files.

Reply:
The runtime includes the files at compile time, so they end up in the data section of the binary. It essentially makes these files part of the code, not a config file that can be changed independently of the code. This makes sense because everybody must use the same values here, across mainnet, testnet, and various testing setups. Nobody should be changing those values dynamically.

The biggest benefit of the approach is that one can run the binary without needing the files. In contrast, config.json is read at runtime, so the node maintainer has to have a copy of it, which they can chose to optimize how they want to run the node.

You are reading files at runtime, so it's more like config.json. This is at a static path, too.

Adding a CLI option might be overkill here and I can see how it would make ugly code changes. I think it would be fine to add a static path, but please make sure it's not at a patch like `"core/chain-configs/res/"` which assumes we are in a developer environment where the full nearcore repository is checked out.

Another option would be to add a line to `config.json` and specify it that way.

cc @nikurt, I think the node team should have the last say about how this is best specified

Reply:
Sure, I will use the input from @nikurt or someone from the team regarding this. My preference from my current understanding is adding the path to the `config.json`

Reply:
Without looking at the code, my preference is the following with some reasonable default. If that default file is missing, handle it gracefully (ignore completely).

> add a line to config.json and specify it that way.